Update BlockStylesSelector component and implement in Eventbrite

scruffian opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

This makes some changes to the BlockStylesSelector component and implements it in the Eventbrite block.

Fixes #14527

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* The preview for the button Embed Type on the EventBrite block will now be a real block preview.
* The EventBrite block also now allows you to change the Embed Type from the BlockControls.
* The `useModal` attribute in the block is now deprecated and replaced with a `style` attribute.
* The component now also includes the inspector controls and creates a the block toolbar

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* adding to existing features

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Add an Eventbrite block
* Add an event
* Check that you can still toggle between a button and an inline embed
<img width="290" alt="Screenshot 2020-01-31 at 13 47 00" src="https://user-images.githubusercontent.com/275961/73544273-bdb38800-4430-11ea-8247-e8d79bd0d636.png">

* Check that the Calendly, OpenTable and EventBrite blocks all have a Toolbar which allows you to change the embed style:
<img width="290" alt="Screenshot 2020-01-31 at 14 16 42" src="https://user-images.githubusercontent.com/275961/73546040-69120c00-4434-11ea-88cd-fba3a82f54b0.png">

* You also need to test the deprecation for old Eventbrite blocks using the old attributes.
* Create an Eventbrite block using master
* Set the Embed type to be "Modal"
* Save the post
* Check out this branch
* Reload the editor
* Check that the block is still in modal type
* Open the editor in code view and check you see `"style":"modal"` in the block attributes.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* no changelog

creativecoder committed:
Eventbrite Block: Update modal embed preview

scruffian committed:
Use the block style selector in the eventbrite block

move the inspector controls and panel body wrappers inside the component

creativecoder commented:
While I appreciate the code consolidation, I think an image is a better fit for the Eventbrite in page embed preview

- The preview is small enough that it's hard to read the text, so there's not really a benefit to having "live" updates
- We don't have to worry about maintaining an external event that could expire or be deleted and break the preview
- There's a significant delay for the embed to render, long enough that it feels like it might be broken when it first loads

The Button & Modal embed does benefit from a live preview, because it shows the button with the style and color defaults of the active theme.

Two ideas to reconcile this

- Modify the `BlockStylesSelector` to accept a customized preview (like an image), OR
- Expose a more granular component, like `BlockStylesSelectorItem`, that can be used for just one of the block styles

Then we could keep the image for the In Page preview and have a dynamic preview for the Button & Modal. What do you think?

scruffian commented:
> We don't have to worry about maintaining an external event that could expire or be deleted and break the preview

I'm not sure what you mean by this, sorry

creativecoder commented:
> We don't have to worry about maintaining an external event that could expire or be deleted and break the preview

I mean the [example url](https://github.com/Automattic/jetpack/pull/14528/files#diff-b6984996c1b052c6449cdc5e6441c04aR98) for the block. If the Eventbrite event at that url is in the past, deleted, or otherwise not working, the preview for the In Page embed won't show correctly.

scruffian commented:
> I mean the example url for the block. If the Eventbrite event at that url is in the past, deleted, or otherwise not working, the preview for the In Page embed won't show correctly.

Ah yes, good point. I'm also not sure why we use the example attributes, rather than previewing with the actual URL the user enters. That's how the core block works, but it doesn't make a lot of sense to me.

I'll think about it some more...

scruffian commented:
The `BlockStyleSelector` component now renders a preview if one is passed, otherwise it uses the `BlockPreview` component, so I think this is ready for another review.

Copons commented:
I think we have an issue with how we slot-fill the various block controls, that is causing the toolbar controls to never show up.

The way we are using the `BlockStylesSelector` in the various blocks is inside the `InspectorControls` slot:
```jsx
<InspectorControls>
  <PanelBody>
    <BlockStylesSelector />
  </PanelBody>
</InspectorControls>
```

But then, inside of `BlockStylesSelector` we also define the Style toolbar button, which is supposed to be rendered inside the `BlockControls` slot:
```jsx
<BlockControls>
  <Toolbar />
</BlockControls>
```

In my tests, the Style button never shows up in the block toolbar.
I've tried moving a bunch of stuff around, and it turns out we could solve the issue and DRY-fy this a bit more.

```jsx
// block edit.js
<BlockStylesSelector />

// BlockStylesSelector
<BlockControls>
  <Toolbar />
</BlockControls>

<InspectorControls>
  <PanelBody>
    <div className="block-editor-block-styles jetpack-block-styles-selector">
      { /* ... */ }
    </div>
  </PanelBody>
</InspectorControls>
```

scruffian commented:
> In my tests, the Style button never shows up in the block toolbar.

Interesting, it shows for me:
<img width="283" alt="Screenshot 2020-02-14 at 15 06 08" src="https://user-images.githubusercontent.com/275961/74542461-933af200-4f3b-11ea-9b57-66d84bad93a8.png">

If removing the nesting fixes it for you though then we might as well do it. Maybe you're running an older version of Gutenberg?

Copons commented:
> Maybe you're running an older version of Gutenberg?

That was it.
I was testing with whatever version is included in WP 5.3.2.
I've tried with Guten plugin 7.4, and indeed the button shows up regardless.

> If removing the nesting fixes it for you though then we might as well do it.

Yeah, it would be a nice compatibility fix, while also helping to keep things simple. 🙂

Copons commented on extensions/blocks/calendly/edit.js:
```diff
@@ -185,18 +185,14 @@ export default function CalendlyEdit( { attributes, className, clientId, setAttr
-						/>
-					</PanelBody>
-				</>
+				<BlockStylesSelector
```
I think this should go outside the `<InspectorControls>` right above.

scruffian commented:
I added a commit to get this working in Wordpress 5.3.

I also noticed an issue with the Eventbrite block where the "styles" panel was appearing in the inline mode because the preview was triggering it. I added an `isSelected` check to where we show the style controls, to get around this.

Copons commented on extensions/blocks/eventbrite/index.js:
```diff
@@ -51,8 +51,8 @@ export const settings = {
 		},
-		useModal: {
-			type: 'boolean',
+		style: {
```
I think this would break existing blocks using the `useModal` attribute.
We should provide a deprecation/migration path: https://developer.wordpress.org/block-editor/developers/block-api/block-deprecation/

scruffian replied:
Good point, I added a `deprecated` array. Thanks :)

Copons commented on extensions/blocks/opentable/edit.js:
```diff
@@ -224,12 +200,7 @@ export default function OpenTableEdit( { attributes, setAttributes, className, c
-					{ blockControls }
-				</>
-			) }
+			{ ! isEmpty( rid ) && <>{ inspectorControls() }</> }
```
We shouldn't need the empty fragment anymore.

Copons commented:
This is working well for me! ✨ 

I've left a nitpick, but mostly I'd love to see the PR title and description updated with the fact that this is also updating Calendly and OpenTable. 🙂

scruffian commented:
> I'd love to see the PR title and description updated with the fact that this is also updating Calendly and OpenTable. 🙂

Done!

scruffian committed:
simplify code

jeherve reviewed (changes requested):
This is going to need a rebase now, I am afraid.

pablinos committed:
Ensure InspectorControls aren't displayed with an invalid URL

When an Eventbrite URL couldn't be embedded the inspectorControls would
still display as the URL prop would have a value. This swaps the check
to the `cannotEmbed` method.

pablinos commented:
> I added a commit to get this working in Wordpress 5.3.

Thought, I'd better check it in 5.2 as well, and it works fine

scruffian commented:
This is rebased and ready for a (final?!) review.

Copons reviewed (dismissed):
This seems to be working fine for me! ✨

jeherve reviewed (commented):
Right now, old Eventbrite blocks seem to display both the link and the embed instead of just one, in my tests:

**editor**
![image](https://user-images.githubusercontent.com/426388/75196555-89479900-575c-11ea-95f6-910888b2803a.png)

**frontend**
![image](https://user-images.githubusercontent.com/426388/75196577-96fd1e80-575c-11ea-9c8a-42ce729bf38a.png)

Do you experience similar issues?

kraftbj commented:
Coming into this PR late, so forgive me if ill-timed. When switching back and forth between button+modal back to inline, there is enough of a pause without any UI indicators where I wondered if it was broken for a second.

scruffian commented:
> there is enough of a pause without any UI indicators where I wondered if it was broken for a second.

I agree that this is a problem, but it is like this on master too, so I don't think it should stop us merging.

jeherve commented:
I am still running into this issue:
https://github.com/Automattic/jetpack/pull/14528#pullrequestreview-363746857

Am I the only one experiencing this?

pablinos commented:
> I am still running into this issue:
> [#14528 (review)](https://github.com/Automattic/jetpack/pull/14528#pullrequestreview-363746857)
> 
> Am I the only one experiencing this?

@jeherve - I've been trying to reproduce this and can't at the moment. I have had an issue with the modal not working unless I'm on `https` but then the link opens in a new window. Firefox blocked popups for me though, so it might be worth investigating that as a separate issue.

Just to confirm that I'm testing correctly. I'm creating an Eventbrite block with the current master branch and setting it to the button style. I then switch to this branch and view it in the frontend. I also edited the post in case a transform was causing it, but everything seems to be working correctly.

Can you reliably reproduce this? I'd love to get to the bottom of it.

pablinos committed:
Check the new and old attribute for whether to use the modal style.

pablinos commented:
> Can you reliably reproduce this? I'd love to get to the bottom of it.

Strike that! I managed to see it. I think I was having some internet issues and I wasn't seeing the inline version straightaway.

Hopefully it's fixed with 3648ee9fe

scruffian commented:
Updating the labels since @pablinos has fixed this

apeatling reviewed (dismissed):
I went through all instructions and confirmed this is working well. Switching from master and creating the block in modal view, then switching to this branch worked correctly. I saw the style: modal in the code view and it remained in the modal style.

apeatling commented:
![](https://user-images.githubusercontent.com/275961/73546040-69120c00-4434-11ea-88cd-fba3a82f54b0.png)

One minor non blocker. Looking at this menu, is it expecting icons? The left edge padding seems too little.

scruffian commented:
> Looking at this menu, is it expecting icons? 

Yeah, but I couldn't think of any. If you have ideas, feel free to add them :)

Copons commented:
I agree with @apeatling about the icons, but I don't think there are good fits among the Dashicons.

This PR is just a one-liner shy from being icon-ready: basically just add `icon: styleOption.icon` in the component's Toolbar, and then the consumer blocks would just need to add whatever icon in their style options arrays.
We can either add it now, or leave to whenever we figure out what icons (custom SVGs?) to use for our blocks.
Either way, I don't think this should block the PR! :shipit:

jeherve reviewed (changes requested):
This is going to need a rebase now, I am afraid.

Copons commented:
> This is going to need a rebase now, I am afraid.

It should be a straightforward one at least: just remove the import of `Toolbar` and add `withNotices` in the two conflicting files.

scruffian commented:
rebase ✅

jeherve commented on extensions/blocks/eventbrite/eventbrite.php:
```diff
@@ -41,7 +41,7 @@ function jetpack_render_eventbrite_block( $attr, $content ) {
 
 	// Show the embedded version.
-	if ( empty( $attr['useModal'] ) ) {
+	if ( empty( $attr['useModal'] ) && 'modal' !== $attr['style'] ) {
```
This triggers a notice when the style is not set:

```
Notice: Undefined index: style in wp-content/plugins/jetpack/extensions/blocks/eventbrite/eventbrite.php on line 44
```

scruffian replied:
Fixed in b1d75b0

jeherve commented on extensions/blocks/eventbrite/deprecated.js:
```diff
@@ -0,0 +1,147 @@
+import { _x } from '@wordpress/i18n';
```
Do you think you could move this into a `eventbrite/deprecated/v1/save.js`, to stay consistent with how we do it for other blocks?

scruffian replied:
Done in adeb0ad

apeatling reviewed (dismissed):
Ran through steps again and all working well.

scruffian committed:
move the deprecation to a directory

scruffian committed:
fix PHP notice

Copons commented:
This works fine, but there seems to be some visual inconsistencies with the latest Gutenberg (7.7, but checked out from [master right now](https://github.com/WordPress/gutenberg/tree/3848806ccc7b8a65778366e80af7ea63258538d4)), that doesn't happen without the Gutenberg plugin.

| Block | No Plugin | Gutenberg 7.7 |
| - | - | - |
| Calendly | <img width="267" alt="Screenshot 2020-03-12 at 12 31 59" src="https://user-images.githubusercontent.com/2070010/76522006-bdf66880-645d-11ea-9c2a-8d3693ac3943.png"> | <img width="265" alt="Screenshot 2020-03-12 at 12 30 20" src="https://user-images.githubusercontent.com/2070010/76522021-c51d7680-645d-11ea-9c8d-c49fe338d16b.png"> |
| Eventbrite | <img width="267" alt="Screenshot 2020-03-12 at 12 32 03" src="https://user-images.githubusercontent.com/2070010/76522056-d5355600-645d-11ea-8bbc-7f5721eab304.png"> | <img width="263" alt="Screenshot 2020-03-12 at 12 30 26" src="https://user-images.githubusercontent.com/2070010/76522069-d9617380-645d-11ea-9818-a46b2f7ccd91.png"> |
| OpenTable | <img width="262" alt="Screenshot 2020-03-12 at 12 32 10" src="https://user-images.githubusercontent.com/2070010/76522089-e1b9ae80-645d-11ea-8fd9-d5315efb2de8.png"> | <img width="267" alt="Screenshot 2020-03-12 at 12 31 13" src="https://user-images.githubusercontent.com/2070010/76522206-1e85a580-645e-11ea-8ddb-a69f706baff3.png"> |

scruffian commented:
Thanks for finding this @Copons. This will be an issue even if we don't merge this PR, so I think we should go ahead and merge and then fix in a different issue.

Copons reviewed (approved):
> This will be an issue even if we don't merge this PR, so I think we should go ahead and merge and then fix in a different issue.

Makes sense @scruffian, go for it!

jeherve reviewed (approved):
This looks good to me. I agree that the inconsistencies between Gutenberg versions aren't ideal though, but we can address this in a follow-up PR.

creativecoder commented:
> This works fine, but there seems to be some visual inconsistencies with the latest Gutenberg (7.7, but checked out from master right now), that doesn't happen without the Gutenberg plugin.

@Copons Thanks for finding these! Can you please open a Jetpack issue for these (if there isn't one already) and add to our global backlog?

scruffian commented:
r204888-wpcom
